### PR TITLE
Implementation of getSearchConditions() 

### PR DESCRIPTION
**Pull request by ZMarouani**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 1]
```diff
@@ -18,14 +18,33 @@
 
 package org.kopi.galite.form
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:19]: merge class with master

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 2]
```diff
@@ -1720,8 +1743,31 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
   /**
    * Returns the search conditions for database query.
    */
-  fun getSearchConditions(): String? {
-    TODO()
+  fun getSearchConditions(): Op<Boolean> {
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1746]: i think that this method can return null value

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 3]
```diff
@@ -1720,8 +1743,31 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+    val conditionList : MutableList<Op<Boolean>> = mutableListOf()
+
+    fields.forEach { vField ->
+      if(vField.getColumnCount() > 0 ){
+        val cond = vField.getSearchCondition_()
+
+        cond?.let {
+          val condOperator : (ExpressionWithColumnType<String>.(Any) -> Op<Boolean>) = vField.getSearchCondition_()!!.first
+          val condOperand : TypeVariable<GenericDeclaration> = vField.getSearchCondition_()!!.second
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1755]: change condOperand to value

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 4]
```diff
@@ -1720,8 +1743,31 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+          val condColumn = Column<String>(tables!![vField.getColumn(0)!!.getTable()],
+                  vField.getColumn(0)!!.getQualifiedName() ,
+                  VarCharColumnType())
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1758]: fix style

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 5]
```diff
@@ -652,6 +656,10 @@ abstract class VField protected constructor(width: Int, height: Int) : VConstant
     }
   }
 
+  open fun getSearchCondition_() : Pair<ExpressionWithColumnType<String>.(t: Any)-> Op<Boolean> , TypeVariable<GenericDeclaration>>? {
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:659]: fix code style

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 6]
```diff
@@ -1720,8 +1743,31 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
   /**
    * Returns the search conditions for database query.
    */
-  fun getSearchConditions(): String? {
-    TODO()
+  fun getSearchConditions(): Op<Boolean> {
+    val conditionList : MutableList<Op<Boolean>> = mutableListOf()
+
+    fields.forEach { vField ->
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1749]: change vField to field

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 7]
```diff
@@ -1720,8 +1743,31 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+      if(vField.getColumnCount() > 0 ){
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1750]: add empty space after if an before {

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt (hunk 3 above)
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1755]: remove : type

**Review by h-haddad — APPROVED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 8]
```diff
@@ -1199,7 +1201,8 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
 
     while (i >= 0) {
       if (listeners[i] == BlockRecordListener::class.java) {
-        (listeners[i + 1] as BlockRecordListener).blockRecordChanged(getSortedPosition(record - 1) + 1, localRecordCount)
+        (listeners[i + 1] as BlockRecordListener).blockRecordChanged(getSortedPosition(record - 1) + 1,
+                                                                     localRecordCount)
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1205]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 9]
```diff
@@ -1807,7 +1837,7 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
 
         if (query.toList().isEmpty()) {
           throw VExecFailedException(MessageCode.getMessage("VIS-00016",
-                  arrayOf<Any>(tables!![tableIndex])))
+                                                            arrayOf<Any>(tables!![tableIndex])))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1840]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 10]
```diff
@@ -1821,7 +1851,7 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
           if (query.toList().isNotEmpty()) {
 
             throw VExecFailedException(MessageCode.getMessage("VIS-00020",
-                    arrayOf<Any>(tables!![tableIndex])))
+                                                              arrayOf<Any>(tables!![tableIndex])))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1854]: undo all reformat changes

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 11]
```diff
@@ -1203,7 +1204,7 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
     while (i >= 0) {
       if (listeners[i] == BlockRecordListener::class.java) {
         (listeners[i + 1] as BlockRecordListener).blockRecordChanged(getSortedPosition(record - 1) + 1,
-                                                                     localRecordCount)
+                localRecordCount)
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1207]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 12]
```diff
@@ -1359,7 +1360,7 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
    */
   protected fun clearRecordImpl(recno: Int) {
     assert(this !== form.getActiveBlock() || isMulti() && recno != activeRecord
-                   || !isMulti() && activeField == null) {
+            || !isMulti() && activeField == null) {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1363]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 13]
```diff
@@ -1727,8 +1728,7 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
    * Returns the tables for database query, with outer joins conditions.
    */
   fun getSearchTables_(): Join? {
-    return VBlockDefaultOuterJoin.getSearchTables(this)
-  }
+    return VBlockDefaultOuterJoin.getSearchTables(this)  }
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1731]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 14]
```diff
@@ -2552,7 +2550,7 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
    * @param     block           This action should block the UI thread ?
    */
   @Deprecated("Use method performAsyncAction without bool parameter",
-              ReplaceWith("performAsyncAction(action)"))
+          ReplaceWith("performAsyncAction(action)"))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:2553]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 15]
```diff
@@ -2979,11 +2977,11 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
-                       title,
-                       this.help,
-                       commands,
-                       fields,
-                       form.blocks.size == 1)
+              title,
+              this.help,
+              commands,
+              fields,
+              form.blocks.size == 1)
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:2984]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 16]
```diff
@@ -1985,10 +2069,10 @@ abstract class VField protected constructor(width: Int, height: Int) : VConstant
...
       lab = lab.replace(' ', '_')
       help.helpOnField(block!!.title,
-                       block!!.getFieldPos(this),
-                       label,
-                       lab ?: name,
-                       toolTip)
+              block!!.getFieldPos(this),
+              label,
+              lab ?: name,
+              toolTip)
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:2075]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 17]
```diff
@@ -2043,10 +2127,10 @@ abstract class VField protected constructor(width: Int, height: Int) : VConstant
...
     }
     help.helpOnType(modeName,
-                    modeDesc,
-                    getTypeName(),
-                    getTypeInformation(),
-                    names)
+            modeDesc,
+            getTypeName(),
+            getTypeInformation(),
+            names)
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:2133]: undo

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 18]
```diff
@@ -2358,7 +2442,7 @@ abstract class VField protected constructor(width: Int, height: Int) : VConstant
    * It is the first line of the field help
    * @return    the help of this field
    */
-  var toolTip: String? = null // help text
+  var toolTip : String? = null // help text
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:2445]: undo

**Review by h-haddad — CHANGES_REQUESTED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt [hunk 19]
```diff
@@ -1741,36 +1742,32 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
...
+          VConstants.FDO_SEARCH_NONE -> condColumn
+          VConstants.FDO_SEARCH_UPPER -> condColumn.upperCase()
+          VConstants.FDO_SEARCH_LOWER -> condColumn.lowerCase()
+          else -> throw InconsistencyException("FATAL ERROR: bad search code: $options")
+        }
+
+        val condition = field.getSearchCondition_(searchColumn)
+
+        condition?.let {
+          conditionList.add(condition)
```
[the remark this answers is not in the document]
Comment (ZMarouani) [reply galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt:1941]: This class is responsible for tables outer joins which is already supported in Exposed  when we use table.join()

**Review by h-haddad — APPROVED** [review]

**Pull request merged**